bigquery/storage_session.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bigquery
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	gax "github.com/googleapis/gax-go/v2"
+	"google.golang.org/api/iterator"
+	bqStoragepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
+	storagepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
+	"google.golang.org/grpc"
+)
+
+// ReadSession is the abstraction over a storage API read session.
+type ReadSession struct {
+	settings *settings
+	rc       *ReadClient
+
+	ctx     context.Context
+	table   *Table
+	tableID string
+
+	bqSession *bqStoragepb.ReadSession
+
+	// SerializedArrowSchema is an IPC-serialized Arrow Schema.
+	SerializedArrowSchema []byte
+
+	// EstimatedTotalBytesScanned on the number of bytes this session will scan when
+	// all streams are completely consumed. This estimate is based on
+	// metadata from the table which might be incomplete or stale.
+	EstimatedTotalBytesScanned int64
+
+	// StreamCount represents the number of streams opened to for this session.
+	// Available after session is initialized.
+	StreamCount int
+
+	// ReadStreams contains at least one stream that is created with
+	// given the session, in the form
+	// projects/{project_id}/locations/{location}/sessions/{session_id}/streams/{stream_id}.
+	// Available after session is initialized.
+	ReadStreams []string
+
+	// SessionID is a unique identifier for the session, in the form
+	// projects/{project_id}/locations/{location}/sessions/{session_id}.
+	// Available after session is initialized.
+	SessionID string
+
+	// ExpireTime at which the session becomes invalid. After this time, subsequent
+	// requests to read this Session will return errors.
+	ExpireTime *time.Time
+}
+
+type settings struct {
+	// MaxStreamCount governs how many parallel streams
+	// can be opened.
+	MaxStreamCount int
+}
+
+func defaultSettings() *settings {
+	return &settings{
+		MaxStreamCount: 0,
+	}
+}
+
+// Run initiates a read session
```

Comment:
Should we instead expose a `NewReadSession()` that handles all this setup as part of instantiation?

Reply:
I added this `Run` method so if users want to lazy initialize the session they can, similarly on how our Query struct works, that users can call `Run` and get a `Job` and `Read` it later, or just call `query.Read` directly. In this case here, it users do not call run, we initialise it on the first `Next` call